README
NAME
    DateTime::Format::Flexible - DateTime::Format::Flexible - Flexibly parse
    strings and turn them into DateTime objects.

SYNOPSIS
      use DateTime::Format::Flexible;
      my $dt = DateTime::Format::Flexible->build( 'January 8, 1999' );
      # $dt = a DateTime object set at 1999-01-08T00:00:00

DESCRIPTION
    If you have ever had to use a program that made you type in the date a
    certain way and thought "Why can't the computer just figure out what
    date I wanted?", this module is for you.

    It attempts to take any string you give it and parse it into a DateTime
    object.

    For supported string formats, see the test file. If you can think of any
    that I do not cover, please let me know.

USAGE
    This module uses DateTime::Format::Builder under the covers. It only has
    one method (build).

  build
    my $dt = DateTime::Format::Flexible->build( $date );

    A small list of supported formats: YYYYMMDDTHHMMSS YYYYMMDDTHHMM
    YYYYMMDDTHH YYYYMMDD YYYYMM MM-DD-YYYY MM-D-YYYY MM-DD-YY M-DD-YY
    YYYY/DD/MM YYYY/M/DD YYYY/MM/D M-D MM-D M-D-Y Month D, YYYY Mon D, YYYY
    Mon D, YYYY HH:MM:SS ...

    there are 2800+ variations that are detected correctly in the test file.

BUGS
    You cannot use a 1 or 2 digit year as the first field:

     YY-MM-DD # not supported
     Y-MM-DD  # not supported

    It would get confused with MM-DD-YY

    It also prefers the US format of MM-DD over the European DD-MM.

AUTHOR
        Tom Heady
        CPAN ID: thinc
        Punch, Inc.
        cpan@punch.net
        http://www.punch.net/

COPYRIGHT
    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.

SEE ALSO
    perl(1). DateTime::Format::Builder



Changes
Revision history for Perl module DateTime::Format::Flexible

0.26 Tue Apr 22 2014
  - support "MM/DD/YYYY HH:MM" (missing seconds),
  - support "EEE MMM dd HH:mm:ss z yyyy" (UTC timezone inside string)
    - thanks David White rt #94911

0.25 Mon Mar 04 2013
   - support "HH:MM::SS timezone YYYY/MM/DD"
     - thanks Kevin Zwack

0.24 Mon Nov 26 2012
   - fix removal of 'am' in the german language file when we already know what part is the time part
   - fix removal of 'st' when not preceded by a digit
     - thanks Andreas Koenig rt #81432
   - support "YYYYMMDD timezone"

0.23 Thu Jun 14 2012
   - Fix for MM/YYYY (was always setting the base year)
     - thanks John Marling
   - Support MM/YY if MMYY is given as an option
   - pod cleanup

0.22 Mon Jun 11 2012
   - update GPL in the LICENSE file (apparently the FSF has a new address)
     - thanks ppisar rt #74363
   - update POD to match the license file
     - thanks ppisar rt #74358
   - fix typo for saturday in the german language file
     - thanks TMUELLER rt #77721
   - don't check the language extensions if we don't have any non-digits in the string we are checking (excluding delimiters: \/-.:[space])
   - fix typos in the pod documentation, add a link to the german translations
   - add support and more tests for some time-first formats
   - add support for GMT timezone anywhere in the date/time string

0.21 Sun Jan 01 2012
   - handle mm/yyyy and m/yyyy

0.20 Sun Sep 18 2011
   - case insensitive handling of st|nd|rd|th (3rd and 3RD now work)
     - thanks Brett Carson

0.19 Fri Jan 07 2011
   - support for German (de)
     - thanks Mark Trettin
   - better support for DD MM
   - support for natural dates in all languages (3 years ago)

0.18 Sat Jan 01 2011
   - move a test that was hardcoded to a year to the no_year tests.
     - thanks cpantesters

0.17 Tue Oct 26 2010
   - more formats supported
   - support timezone offsets that are not at the end of the datetime string.
     - they must be 4 digits and begin with a plus or minus
     - thanks snarkyboojum: http://use.perl.org/use.perl.org/_snarkyboojum/journal/40297.html
   - better support for dates like 'December 1st'
   - POD formatting fixes
   - support 'Oct.26, 2010'.
     - thanks Brian Knapp

0.16 Thu Aug 25 2010
   - make sure 'now' means now and not when the module loaded unless user has set a base
     - thanks Ryan Voots rt #60731

0.15 Mon Mar 10 2010
   - fix tests for DateTime string overloading problem
     - thanks Andreas Koenig and Michael Schwern

0.14 Sun Feb 28 2010
   - fix test: '1 month ago at 4pm' could be potentially less that 28 days ago.
   - fix tests for infinity, -infinity, infinito, and -infinito

0.13 Sat Feb 27 2010
   - make parsing for infinity and -infinity more reliable

0.12 Thu Feb 25 2010
   - fix DateTime::Format::Builder import

0.11 Wed Feb 24 2010
   - remove Readonly dependency
   - support some postgresql datetimes: epoch, infinity, -infinity, allballs
     - see http://www.postgresql.org/docs/8.4/interactive/datatype-datetime.html section 8.5.1.4
     - 'infinity' returns a DateTime::Infinite::Future object
     - '-infinity' returns a DateTime::Infinite::Past object

0.10 Wed Feb 24 18:45:00 2009
   - support for single character am/pm strings '3p'
   - support for days of the week (wednesday => the nearest future wednesday)
     - thanks Shawn Moore, rt #53188
     - note: this is quite limited, I need more test cases
       - works: wednesday
       - works: wed at 3p
   - support for bare months (february)
   - support for limited timezones that are not at the end of the string
     - thanks Dave Faraldo
     - 'Wed Nov 11 13:55:48 PST 2009' becomes
       - 2009-11-11T13:55:48 America/Los_Angeles
   - beginning multiple language support (contributions welcome)
     - support english (en) and spanish (es)
   - support dates like 'now, today, tomorrow'
   - switch to strptime from regexes for some parsing for clarity
   - now supports negative timezone offsets IF it is a 4 digit offset and there is a space before the offset
     - works: 2007-05-06T04:44:44 -0800
     - does not work: 2007-05-06T04:44:44-0800
     - does not work: 2007-05-06T04:44:44-08
     - does not work: 2007-05-06T04:44:44 -08
   - now supports 2 digit years as the first number if the year is > 31
     - works: 35-12-23 (2035-12-23T00:00:00)
     - does not work: 11-12-13 (2013-11-12T00:00:00, or 2013-12-11T00:00:00 with european hinting)
   - timezone parsing should now be more reliable
   - you can now set a 'base' datetime object to help fill out partial datetimes

0.09 Sun May 24 22:00:00 2009
   - add a real copyright for Debian packaging
   - fix parsing a bare 4 digit year.
     - thanks Dominic Rose, rt #46278

0.08 Wed Apr 22 14:00:00 2009
    - fix some pod errors, give an example for european hinting

0.07 Tue Apr 22 03:00:00 2009
    - add ability to strip strings from the date string
    - can now parse timezones
    - rudimentary support for european dates (dd-mm-yyyy)
    - can now parse epoch times

0.06 Mon Apr 20 15:06:00 2009
    - Fix parsing bug with dd-(oct|nov|dec)-yyyy
      - thanks Phil Brass

0.05 Fri Apr 4 14:00:00 2008
    - fix some test helper libraries not being included.
      - thanks GOOZBACH, rt #34683

0.04 Fri Mar 14 21:00:00 2008
    - fix broken 0.03 release
    - split tests out for easier integration into other distributions

0.03 Mon Mar 10 02:00:00 2008
   - add method "parse_datetime" as an alias for "build".
     - thanks Adriano Rodrigues Ferreira

0.02 Tue Oct 02 01:50:00 2007
   - fixed pod formatting
   - added new parseable format: 2007-10-01 13:11:32.741804.
     - thanks RKRIMEN

0.01 Sun Jul 15 20:09:47 2007
    - original version; created by ExtUtils::ModuleMaker 0.49


t/lib/helper.pm
package t::lib::helper;

use strict;
use warnings;

use Test::More;

use DateTime::Format::Flexible;

my $base = 'DateTime::Format::Flexible';

sub run_tests
{
    my $opts = [];
    if ( ref( $_[0] ) eq 'ARRAY' )
    {
        $opts = shift @_;
    }
    foreach ( @_ )
    {
        my ( $line ) = $_ =~ m{([^\n]+)};
        next if not $line;
        next if $line =~ m{\A\#}mx; # skip comments
        next if $line =~ m{\A\z}mx; # skip blank lines
        my ( $given , $wanted , $tz ) = split m{\s+=>\s+}mx , $line;
        compare( $given , $wanted , $tz, $opts );
    }
}

sub compare
{
    my ( $given , $wanted , $tz, $opts ) = @_;
    my $dt = $base->parse_datetime( $given, @$opts );
    is( $dt->datetime , $wanted , "$given => $wanted" );
    if ( $tz )
    {
        is( $dt->time_zone->name , $tz , "timezone => $tz" );
    }
}

1;


lib/DateTime/Format/Flexible/lang.pm
package DateTime::Format::Flexible::lang;

use strict;
use warnings;

use Module::Pluggable require => 1 , search_path => [__PACKAGE__];
use List::MoreUtils 'any';

sub new
{
    my ( $class , %params ) = @_;
    my $self = bless \%params , $class;
    return $self;
}

sub _cleanup
{
    my ( $self , $date , $p ) = @_;
    foreach my $plug ( $self->plugins )
    {
        if ( $self->{lang} )
        {
            my ( $lang ) = $plug =~ m{(\w{2}\z)}mx;
            if ( not any { $_ eq $lang } @{ $self->{lang} } )
            {
                printf( "# skipping %s\n", $plug ) if $ENV{DFF_DEBUG};
                next;
            }
        }
        printf( "# not skipping %s\n", $plug ) if $ENV{DFF_DEBUG};

        printf( "#   before math: %s\n", $date ) if $ENV{DFF_DEBUG};;
        $date = $self->_do_math( $plug , $date );
        printf( "#   before string_dates: %s\n", $date ) if $ENV{DFF_DEBUG};;
        $date = $self->_string_dates( $plug , $date );
        printf( "#   before fix_alpha_month: %s\n", $date ) if $ENV{DFF_DEBUG};;
        ( $date , $p ) = $self->_fix_alpha_month( $plug , $date , $p );
        printf( "#   before remove_day_names: %s\n", $date ) if $ENV{DFF_DEBUG};;
        $date = $self->_remove_day_names( $plug , $date );
        printf( "#   before fix_hours: %s\n", $date ) if $ENV{DFF_DEBUG};;
        $date = $self->_fix_hours( $plug , $date );
        printf( "#   before remove_strings: %s\n", $date ) if $ENV{DFF_DEBUG};;
        $date = $self->_remove_strings( $plug , $date );
        printf( "#   before locate_time: %s\n", $date ) if $ENV{DFF_DEBUG};;
        $date = $self->_locate_time( $plug , $date );
        printf( "#   before fix_internal_tz: %s\n", $date ) if $ENV{DFF_DEBUG};;
        ( $date , $p ) = $self->_fix_internal_tz( $plug , $date , $p );
        printf( "#   finished: %s\n", $date ) if $ENV{DFF_DEBUG};;
    }
    return ( $date , $p );
}

sub _fix_internal_tz
{
    my ( $self , $plug , $date , $p ) = @_;
    my %tzs = $plug->timezone_map;
    while( my( $orig_tz , $new_tz ) = each ( %tzs ) )
    {
        if( $date =~ m{$orig_tz}mxi )
        {
            $p->{ time_zone } = $new_tz;
            $date =~ s{$orig_tz}{}mxi;
            return ( $date , $p );
        }
    }
    return ( $date , $p );
}

sub _do_math
{
    my ( $self , $plug , $date ) = @_;
    my %strings = $plug->math_strings;
    my $ago = $plug->ago;

    if ( $date =~ m{$ago}mix )
    {
        my $base_dt = DateTime::Format::Flexible->base->clone;
        if ( my ( $amount , $unit ) = $date =~ m{(\d+)\s+([^\s]+)}mx )
        {

            printf( "#    %s => %s\n", $amount, $unit ) if $ENV{DFF_DEBUG};
            if ( exists( $strings{$unit} ) )
            {
                printf( "#    found: %s\n", $strings{$unit} ) if $ENV{DFF_DEBUG};
                my $ret = $base_dt->subtract( $strings{$unit} => $amount );
                $date =~ s{\s{0,}$amount\s+$unit\s{0,}}{}mx;
                printf( "#    after removing amount,unit: [%s]\n", $date ) if $ENV{DFF_DEBUG};
                $date =~ s{$ago}{}mx;
                printf( "#    after removing ago: [%s]\n", $date ) if $ENV{DFF_DEBUG};
                if ( $date ) # we still have more to parse...
                {
                    $date = $ret->ymd . ' ' . $date;
                }
                else
                {
                    $date = $ret->datetime;
                }
                return $date;
            }
        }
    }
    return $date;
}

sub _string_dates
{
    my ( $self , $plug , $date ) = @_;
    my %strings = $plug->string_dates;
    foreach my $key ( keys %strings )
    {
        if ( $date =~ m{\Q$key\E}mxi )
        {
            my $new_value = $strings{$key}->();
            $date =~ s{\Q$key\E}{$new_value}mix;
        }
    }

    my %day_numbers = $plug->day_numbers;
    foreach my $key ( keys %day_numbers )
    {
        if ( $date =~ m{$key}mxi )
        {
            my $new_value = $day_numbers{$key};
            $date =~ s{$key}{n${new_value}n}mix;
        }
    }
    return $date;
}

# turn month names into month numbers with surrounding X
# Sep => X9X
sub _fix_alpha_month
{
    my ( $self , $plug , $date , $p ) = @_;
    my %months = $plug->months;
    while( my( $month_name , $month_number ) = each ( %months ) )
    {
        if( $date =~ m{\b$month_name\b}mxi )
        {
            $p->{ month } = $month_number;
            $date =~ s{\b$month_name\b}{X${month_number}X}mxi;

            return ( $date , $p );
        }
        elsif ( $date =~ m{\d$month_name}mxi )
        {
            $p->{ month } = $month_number;
            $date =~ s{(\d)$month_name}{$1X${month_number}X}mxi;

            return ( $date , $p );
        }
    }
    return ( $date , $p );
}

# remove any day names, we do not need them
sub _remove_day_names
{
    my ( $self , $plug , $date ) = @_;
    my %days = $plug->days;
    foreach my $day_name ( keys %days )
    {
        # if the day name is by itself, make it the upcoming day
        # eg: monday = next monday
        if ( $date =~ m{\A$day_name\z}mx or
             $date =~ m{$day_name\sat}mx )
        {
            my $dt = $self->{base}->clone->truncate( to => 'day' );
            if ( $days{$day_name} == $dt->dow )
            {
                my $str = $dt->ymd;
                $date =~ s{$day_name}{$str};
                return $date;
            }
            elsif ( $days{$day_name} > $dt->dow )
            {
                $dt->add( days => $days{$day_name} - $dt->dow );
                my $str = $dt->ymd;
                $date =~ s{$day_name}{$str};
                return $date;
            }
            else
            {
                $dt->add( days => $days{$day_name} - $dt->dow + 7 );
                my $str = $dt->ymd;
                $date =~ s{$day_name}{$str};
                return $date;
            }
        }
        # otherwise, just strip it out
        if ( $date =~ m{$day_name}mxi )
        {
            $date =~ s{$day_name,?}{}gmix;
            return $date;
        }
    }
    return $date;
}

# fix noon and midnight
sub _fix_hours
{
    my ( $self , $plug , $date ) = @_;
    my %hours = $plug->hours;
    foreach my $hour ( keys %hours )
    {
        if ( $date =~ m{$hour}mxi )
        {
            my $realtime = $hours{ $hour };
            $date =~ s{$hour}{$realtime}gmix;
            return $date;
        }
    }
    return $date;
}

sub _remove_strings
{
    my ( $self , $plug , $date ) = @_;
    my @rs = $plug->remove_strings;
    foreach my $rs ( @rs )
    {
        if ( $date =~ m{$rs}mxi )
        {
            printf( "#     removing string: %s\n", $rs ) if $ENV{DFF_DEBUG};

            $date =~ s{$rs}{ }gmix;
        }
    }
    return $date;
}

sub _locate_time
{
    my ( $self , $plug , $date ) = @_;
    $date = $plug->parse_time( $date );
    return $date;
}

1;

__END__

=head1 NAME

DateTime::Format::Flexible::lang - base language module to handle plugins for DateTime::Format::Flexible.

=head1 DESCRIPTION

You should not need to use this module directly

=head2 new

Instantiate a new instance of this module.

=head1 AUTHOR

    Tom Heady
    CPAN ID: thinc
    Punch, Inc.
    cpan@punch.net
    http://www.punch.net/

=head1 COPYRIGHT & LICENSE

Copyright 2011 Tom Heady.

This program is free software; you can redistribute it and/or
modify it under the terms of either:

=over 4

=item * the GNU General Public License as published by the Free
    Software Foundation; either version 1, or (at your option) any
    later version, or

=item * the Artistic License version 2.0.

=back

=head1 SEE ALSO

F<DateTime::Format::Flexible>

=cut


lib/DateTime/Format/Flexible/lang/de.pm
package DateTime::Format::Flexible::lang::de;

use strict;
use warnings;

sub new
{
    my ( $class , %params ) = @_;
    my $self = bless \%params , $class;
    return $self;
}

sub months
{
    return (
        qr{Jan(?:uar)?}i        => 1,
        qr{Jän(?:er)?}i         => 1, # Austrian?!
        qr{Feb(?:ruar)?}i       => 2,
        qr{Mär(?:z)?|Maerz}i    => 3,
        qr{Apr(?:il)?}i         => 4,
        qr{Mai}i                => 5,
        qr{Jun(?:i)?}i          => 6,
        qr{Jul(?:i)?}i          => 7,
        qr{Aug(?:ust)?}i        => 8,
        qr{Sep(?:tember)?}i     => 9,
        qr{Okt(?:ober)?}i       => 10,
        qr{Nov(?:ember)?}i      => 11,
        qr{Dez(?:ember)?}i      => 12,
    );
}

sub days
{
    return (
        qr{\bMo(?:ntag)?\b}i             => 1, # Monday
        qr{\bDi(?:enstag)?\b}i           => 2, # Tuesday
        qr{\bMi(?:ttwoch)?\b}i           => 3, # Wednesday
        qr{\bDo(?:nnerstag)?\b}i         => 4, # Thursday
        qr{\bFr(?:eitag)?\b}i            => 5, # Friday
        qr{\bSa(?:mmstag)?\b}i           => 6, # Saturday
        qr{\bSonnabend\b}i               => 6, # Saturday
        qr{\bSo(?:nntag)?\b}i            => 7, # Sunday
    );
}

sub day_numbers
{
    return (
        qr{erster}i                                 => 1, # first
        qr{ersten}i                                 => 1, # first
        qr{zweiter}i                                => 2, # second
        qr{dritter}i                                => 3, # third
        qr{vierter}i                                => 4, # fourth
        qr{fünfter|fuenfter}i                       => 5, # fifth
        qr{sechster}i                               => 6, # sixth
        qr{siebter}i                                => 7, # seventh
        qr{achter}i                                 => 8, # eighth
        qr{neunter}i                                => 9, # ninth
        qr{zehnter}i                                => 10, # tenth
        qr{elfter}i                                 => 11, # eleventh
        qr{zwölfter|zwoelfter}i                     => 12, # twelfth
        qr{dreizehnter}i                            => 13, # thirteenth
        qr{vierzehnter}i                            => 14, # fourteenth
        qr{vierzehnten}i                            => 14, # fourteenth
        qr{fünfzehnter|fuenfzehnter}i               => 15, # fifteenth
        qr{sechzehnter}i                            => 16, # sixteenth
        qr{siebzehnter}i                            => 17, # seventeenth
        qr{achtzehnter}i                            => 18, # eithteenth
        qr{neunzehnter}i                            => 19, # ninteenth
        qr{zwanzigster}i                            => 20, # twentieth
        qr{einundzwanzigster}i                      => 21, # twenty first
        qr{zweiundzwanzigster}i                     => 22, # twenty second
        qr{dreiundzwanzigster}i                     => 23, # twenty third
        qr{vierundzwanzigster}i                     => 24, # twenty fourth
        qr{fünfundzwanzigster|fuenfundzwanzigster}i => 25, # twenty fifth
        qr{sechsundzwanzigster}i                    => 26, # twenty sixth
        qr{siebenundzwanzigster}i                   => 27, # twenty seventh
        qr{achtundzwanzigster}i                     => 28, # twenty eighth
        qr{neunundzwanzigster}i                     => 29, # twenty ninth
        qr{dreißigster|dreissigster}i               => 30, # thirtieth
        qr{einunddreißigster|einunddreissigster}i   => 31, # thirty first
    );
}

sub hours
{
    return (
        Mittag       => '12:00:00', # noon
        mittags      => '12:00:00', # noon
        Mitternacht  => '00:00:00', # midnight
        mitternachts => '00:00:00', # midnight
    );
}

sub remove_strings
{
    return (
        # we want to remove ' am ' only when it does not follow a digit
        # if we just remove ' am ', it removes am/pm designation, losing accuracy
        qr{(?<!\d)\sam\b}i, # remove ' am ' as in '20. Feb am Mittag'
        # we can also remove it if it is at the beginning
        qr{\A\bam\b}i,
        qr{\bum\b}i,        # remove ' um ' as in '20. Feb um Mitternacht'
    );
}

sub parse_time
{
    my ( $self, $date ) = @_;
    return $date;
}

sub string_dates
{
    my $base_dt = DateTime::Format::Flexible->base;
    return (
        jetzt   => sub { return $base_dt->datetime },                                                   # now
        heute   => sub { return $base_dt->clone->truncate( to => 'day' )->ymd } ,                       # today
        morgen  => sub { return $base_dt->clone->truncate( to => 'day' )->add( days => 1 )->ymd },      # tomorrow
        gestern => sub { return $base_dt->clone->truncate( to => 'day' )->subtract( days => 1 )->ymd }, # yesterday
        'übermorgen' => sub { return DateTime->today->add( days => 2 )->ymd },  # overmorrow (the day after tomorrow) don't know if the Umlaut works
        uebermorgen  => sub { return DateTime->today->add( days => 2 )->ymd },   # overmorrow (the day after tomorrow)
        Epoche       => sub { return DateTime->from_epoch( epoch => 0 ) },
        '-unendlich' => sub { return '-infinity' },
        unendlich    => sub { return 'infinity'  },
    );
}

sub ago
{
    return qr{\bvor\b}; # as in 3 years ago
}

sub math_strings
{
    return (
        Jahr    => 'years' ,
        Jahre   => 'years' ,
        Jahren  => 'years' ,
        Monat   => 'months' ,
        Monate  => 'months' ,
        Tag     => 'days' ,
        Tage    => 'days' ,
        Stunde  => 'hours' ,
        Stunden => 'hours' ,
        Minute  => 'minutes' ,
        Minuten => 'minutes' ,
    );
}

sub timezone_map
{
    # http://home.tiscali.nl/~t876506/TZworld.html
    return (
        CET  => 'Europe/Berlin',
        CEST => 'Europe/Berlin',
        MEZ  => 'Europe/Berlin', # German Version: Mitteleuropäische Zeit
        MESZ => 'Europe/Berlin', # Mitteleuropäische Sommerzeit
    );
}

1;
__END__

=encoding utf-8

=head1 NAME

DateTime::Format::Flexible::lang::de - german language plugin

=head1 DESCRIPTION

You should not need to use this module directly.

If you only want to use one language, specify the lang property when parsing a date.

example:

 my $dt = DateTime::Format::Flexible->parse_datetime(
     'Montag, 6. Dez 2010' ,
     lang => ['de']
 );
 # $dt is now 2010-12-06T00:00:00

Note that this is not required, by default ALL languages are scanned when trying to parse a date.

=head2 new

Instantiate a new instance of this module.

=head2 months

month name regular expressions along with the month numbers (Jan(:?uar)? => 1)

=head2 days

day name regular expressions along the the day numbers (Montag => 1)

=head2 day_numbers

maps day of month names to the corresponding numbers (erster => 01)

=head2 hours

maps hour names to numbers (Mittag => 12:00:00)

=head2 remove_strings

strings to remove from the date (um as in um Mitternacht)

=head2 parse_time

currently does nothing

=head2 string_dates

maps string names to real dates (jetzt => DateTime->now)

=head2 ago

the word use to denote a date in the past (vor 3 Jahren => 3 years ago)

=head2 math_strings

useful strings when doing datetime math

=head2 timezone_map

maps unofficial timezones to official timezones for this language (MEZ => Europe/Berlin)

=head1 AUTHOR

    Mark Trettin <nulldevice.mark@gmx.de>

    Based on DateTime::Format::Flexible::lang::en by
    Tom Heady
    CPAN ID: thinc
    Punch, Inc.
    cpan@punch.net
    http://www.punch.net/

=head1 COPYRIGHT & LICENSE

Copyright 2011 Mark Trettin.

This program is free software; you can redistribute it and/or
modify it under the terms of either:

=over 4

=item * the GNU General Public License as published by the Free
    Software Foundation; either version 1, or (at your option) any
    later version, or

=item * the Artistic License version 2.0.

=back

=head1 SEE ALSO

F<DateTime::Format::Flexible>

=cut
### Local variables:
### coding: utf-8
### End:


lib/DateTime/Format/Flexible/lang/en.pm
package DateTime::Format::Flexible::lang::en;

use strict;
use warnings;

sub new
{
    my ( $class , %params ) = @_;
    my $self = bless \%params , $class;
    return $self;
}

sub months
{
    return (
        qr{Jan(?:uary)?}i        => 1,
        qr{Feb(?:ruary)?}i       => 2,
        qr{Mar(?:ch)?}i          => 3,
        qr{Apr(?:il)?}i          => 4,
        qr{May}i                 => 5,
        qr{Jun(?:e)?}i           => 6,
        qr{Jul(?:y)?}i           => 7,
        qr{Aug(?:ust)?}i         => 8,
        qr{Sep(?:t)?(?:ember)?}i => 9,
        qr{Oct(?:ober)?}i        => 10,
        qr{Nov(?:ember)?}i       => 11,
        qr{Dec(?:ember)?}i       => 12,
    );
}

sub days
{
    return (
        qr{\bMon(?:day)?\b}i    => 1,
        qr{\bTue(?:sday)?\b}i   => 2,
        qr{\bWed(?:nesday)?\b}i => 3,
        qr{\bThu(?:rsday)?\b}i  => 4,
        qr{\bFri(?:day)?\b}i    => 5,
        qr{\bSat(?:urday)?\b}i  => 6,
        qr{\bSun(?:day)?\b}i    => 7,
    );
}

sub day_numbers
{
    return (
        qr{first}            => 1,
        qr{second}           => 2,
        qr{third}            => 3,
        qr{fourth}           => 4,
        qr{fifth}            => 5,
        qr{sixth}            => 6,
        qr{seventh}          => 7,
        qr{eighth}           => 8,
        qr{ninth}            => 9,
        qr{tenth}            => 10,
        qr{eleventh}         => 11,
        qr{twelfth}          => 12,
        qr{thirteenth}       => 13,
        qr{fourteenth}       => 14,
        qr{fifteenth}        => 15,
        qr{sixteenth}        => 16,
        qr{seventeenth}      => 17,
        qr{eithteenth}       => 18,
        qr{ninteenth}        => 19,
        qr{twentieth}        => 20,
        qr{twenty\s?first}   => 21,
        qr{twenty\s?second}  => 22,
        qr{twenty\s?third}   => 23,
        qr{twenty\s?fourth}  => 24,
        qr{twenty\s?fifth}   => 25,
        qr{twenty\s?sixth}   => 26,
        qr{twenty\s?seventh} => 27,
        qr{twenty\s?eighth}  => 28,
        qr{twenty\s?ninth}   => 29,
        qr{thirtieth}        => 30,
        qr{thirty\s?first}   => 31,
    );
}

sub hours
{
    return (
        noon     => '12:00:00' ,
        midnight => '00:00:00' ,
    );
}

sub remove_strings
{
    return (
        qr{\bof\b}i,              # remove ' of ' as in '16th of November 2003'
        qr{(?:st|nd|rd|th)\b,?}i, # remove number extensions
        qr{\bnext\b}i,            # next sunday
    );
}

sub parse_time
{
    my ( $self, $date ) = @_;

    return $date if ( not $date =~ m{\s?at\s?}mx );
    my ( $pre, $time, $post ) = $date =~ m{\A(.+)\s?at\s?([\d\.:]+)(.+)?\z}mx;
    $post ||= q{};

    $date = $pre . 'T' . $time . 'T' . $post;
    return $date;
}

sub string_dates
{
    my $base_dt = DateTime::Format::Flexible->base;
    return (
        now         => sub { return $base_dt->datetime } ,
        today       => sub { return $base_dt->clone->truncate( to => 'day' )->ymd } ,
        tomorrow    => sub { return $base_dt->clone->truncate( to => 'day' )->add( days => 1 )->ymd },
        yesterday   => sub { return $base_dt->clone->truncate( to => 'day' )->subtract( days => 1 )->ymd },
        overmorrow  => sub { return $base_dt->clone->truncate( to => 'day' )->add( days => 2 )->ymd },
        allballs    => sub { return $base_dt->clone->truncate( to => 'day' ) },

        epoch       => sub { return DateTime->from_epoch( epoch => 0 ) },
        '-infinity' => sub { '-infinity' },
        infinity    => sub { 'infinity'  },
    );
}

sub ago
{
    return qr{\bago\b}; # as in 3 years ago
}

sub math_strings
{
    return (
        year   => 'years' ,
        years  => 'years' ,
        month  => 'months' ,
        months => 'months' ,
        day    => 'days' ,
        days   => 'days' ,
        hour   => 'hours' ,
        hours  => 'hours' ,
        minute => 'minutes' ,
        minutes => 'minutes' ,
    );
}

sub timezone_map
{
    # http://home.tiscali.nl/~t876506/TZworld.html
    return (
        EST => 'America/New_York',
        EDT => 'America/New_York',
        CST => 'America/Chicago',
        CDT => 'America/Chicago',
        MST => 'America/Denver',
        MDT => 'America/Denver',
        PST => 'America/Los_Angeles',
        PDT => 'America/Los_Angeles',
        AKST => 'America/Juneau',
        AKDT => 'America/Juneau',
        HAST => 'America/Adak',
        HADT => 'America/Adak',
        HST => 'Pacific/Honolulu',
    );
}

1;
__END__

=head1 NAME

DateTime::Format::Flexible::lang::en - the english language plugin

=head1 DESCRIPTION

You should not need to use this module directly.

If you only want to use one language, specify the lang property when parsing a date.

example:

 my $dt = DateTime::Format::Flexible->parse_datetime(
     'Wed, Jun 10, 2009' ,
     lang => ['en']
 );
 # $dt is now 2009-06-10T00:00:00

Note that this is not required, by default ALL languages are scanned when trying to parse a date.

=head2 new

Instantiate a new instance of this module.

=head2 months

month name regular expressions along with the month numbers (Jan(?:uary)? => 1)

=head2 days

day name regular expressions along the the day numbers (Mon(?:day)? => 1)

=head2 day_numbers

maps day of month names to the corresponding numbers (first => 01)

=head2 hours

maps hour names to numbers (noon => 12:00:00)

=head2 remove_strings

strings to remove from the date (rd as in 3rd)

=head2 parse_time

searches for the string 'at' to help determine a time substring (sunday at 3:00)

=head2 string_dates

maps string names to real dates (now => DateTime->now)

=head2 ago

the word used to denote a date in the past (3 years ago)

=head2 math_strings

useful strings when doing datetime math

=head2 timezone_map

maps unofficial timezones to official timezones for this language (CST => America/Chicago)


=head1 AUTHOR

    Tom Heady
    CPAN ID: thinc
    Punch, Inc.
    cpan@punch.net
    http://www.punch.net/

=head1 COPYRIGHT & LICENSE

Copyright 2011 Tom Heady.

This program is free software; you can redistribute it and/or
modify it under the terms of either:

=over 4

=item * the GNU General Public License as published by the Free
    Software Foundation; either version 1, or (at your option) any
    later version, or

=item * the Artistic License version 2.0.

=back

=head1 SEE ALSO

F<DateTime::Format::Flexible>

=cut
